cluster-autoscaler/cloudprovider/mcm/fakeclient/client.go

```diff
@@ -0,0 +1,592 @@
+/*
+Copyright (c) 2023 SAP SE or an SAP affiliate company. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fakeclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	fakeuntyped "github.com/gardener/machine-controller-manager/pkg/client/clientset/versioned/fake"
+	apipolicyv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/discovery"
+	fakediscovery "k8s.io/client-go/discovery/fake"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	policyv1beta1 "k8s.io/client-go/kubernetes/typed/policy/v1beta1"
+	fakepolicyv1beta1 "k8s.io/client-go/kubernetes/typed/policy/v1beta1/fake"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+// FakeObjectTracker implements both k8stesting.ObjectTracker as well as watch.Interface.
+type FakeObjectTracker struct {
+	*watch.FakeWatcher
```

Comment:
The `fakeObjectTracker` is used to create the `fakeClient`, which is then used to create the `Informers`. The watcher slice is filled dynamically when the `watchReactionFunc` is called by the informer of the individual resource type. So for each resourceType (Machine, MachineDeployment etc.), there will be a separate watcher; hence, we need a slice of watchers in the `fakeObjectTracker`. The `*watch.FakeWatcher` will create an event on a result channel that the `Start` method of the `fakeObjectTracker` will read from and pass to all the watchers. The watcher which can handle the event will raise an event and populate its own result channel from where the respective controllers will pick the event and reconcile the corresponding object.

Reply:
Ok one more doubt. You call `watchReactorFunc` from 2 places and in both the places you have set the resource to "*" which i presume means all resource types, right? This means there is one registered `watchReactorFunc` only which gets added to the upstream `Fake.WatchReactionChain`. So when `Fake.InvokesWatch` is called then it will just fine one `WatchReactor` for all resources. Therefore your statement: `So for each resourceType (Machine, MachineDeployment etc.), there will be a separate watcher; hence, we need a slice of watchers in the fakeObjectTracker` is not completely understood.